### PR TITLE
feat(runtime): Add per-task ring sizing to RunConfig

### PR DESCRIPTION
**Pull request by lyfne123**
## Summary

The runtime recently added per-task ring sizing via `CallConfig.runtime_env`
(runtime #1042). This wires it through to PyPTO's `RunConfig` so callers can
size the per-task rings per invocation.

- Add three optional overrides to `RunConfig`: `ring_task_window`, `ring_heap`,
  `ring_dep_pool`. `None` (default) leaves the field unset so the runtime falls
  back to its `PTO2_RING_*` env var / compile-time default.
- `_build_call_config` transcribes each into `CallConfig.runtime_env` **only
  when set**, leaving the field at its `0` default otherwise — matching the
  precedence the runtime expects (per-task value > env var > compile-time
  default). Both `CompiledProgram.build_call_config` paths forward `run_config`
  unchanged, so the new fields propagate automatically.
- `RunConfig` validates the values early (power-of-2 ≥ 4 for the task window,
  power-of-2 ≥ 1024 bytes for the heap, `[4, INT32_MAX]` for the dep pool),
  mirroring the runtime's `RuntimeEnv::validate()` so callers get a clear error
  at construction rather than a deep failure inside the runtime.

`RunConfig` is per-invocation, which maps exactly onto the runtime's "per L2
submission" granularity — no IR / pass / codegen changes are required.

## Testing

- [x] Added unit tests in `tests/ut/runtime/test_run_config.py`: field
  defaults, valid/boundary values, the three validation rules, and
  `_build_call_config` transcription (set / unset / partial) via a spy
  `CallConfig` so they run without the optional `simpler` package. 20 new tests
  pass; `ruff` clean.
- [x] Code review completed.

## Related Issues

Adapts PyPTO to runtime per-task ring sizing (`CallConfig.runtime_env`).

### CODE REVIEW

File: python/pypto/runtime/runner.py [hunk 1]
```diff
@@ -228,6 +246,36 @@ def __post_init__(self) -> None:
...
+                f"ring_task_window must be a power of 2 >= 4, got {self.ring_task_window}"
+            )
+        if self.ring_heap is not None and not (_is_pow2(self.ring_heap) and self.ring_heap >= 1024):
+            raise ValueError(
+                f"ring_heap must be a power of 2 >= 1024 (bytes per ring), got {self.ring_heap}"
+            )
+        if self.ring_dep_pool is not None and not (4 <= self.ring_dep_pool <= 2**31 - 1):
+            raise ValueError(
+                f"ring_dep_pool must be in [4, INT32_MAX], got {self.ring_dep_pool}"
+            )
```
[the remark this answers is not in the document]
Comment (lyfne123) [reply python/pypto/runtime/runner.py:279]: Fixed in `25607032`. `_validate_ring_overrides` now rejects non-integers (floats and bools) with a clear `ValueError` via an `_is_int` guard before the power-of-2 bitwise check, so `ring_dep_pool=4.5` / `ring_task_window=True` no longer slip through or raise `TypeError`. Added tests covering float/bool rejection.

File: python/pypto/runtime/runner.py [hunk 2]
```diff
@@ -228,6 +246,36 @@ def __post_init__(self) -> None:
...
+
```
[the remark this answers is not in the document]
Comment (lyfne123) [reply python/pypto/runtime/runner.py:284]: Keeping the early validation intentionally. The runtime's `RuntimeEnv::validate()` only fires at device submission, so codegen-only / sim paths and the `AttributeError` surface (when an older installed `simpler` lacks `runtime_env`) would otherwise fail deep and unclearly. Validating in `RunConfig.__post_init__` gives a clear `ValueError` at the call site. The constraints (pow2>=4, pow2>=1024, [4,INT32_MAX]) are stable hardware/runtime invariants with low drift risk, and there is no PyPTO rule under `.claude/rules/` forbidding cross-layer fail-fast validation. This mirrors the existing fail-fast checks in `RunConfig` (e.g. platform validation).

File: tests/ut/runtime/test_run_config.py [hunk 3]
```diff
@@ -148,6 +148,121 @@ def test_dfx_opts_any_false_when_all_off(self):
...
+    @pytest.mark.parametrize("bad", [512, 1000, 1536, 0])
+    def test_ring_heap_must_be_pow2_ge1024(self, bad):
+        with pytest.raises(ValueError, match="ring_heap must be a power of 2 >= 1024"):
+            RunConfig(platform="a2a3sim", ring_heap=bad)
+
+    @pytest.mark.parametrize("bad", [3, 0, -1, 2**31])
+    def test_ring_dep_pool_must_be_in_int32_range(self, bad):
+        with pytest.raises(ValueError, match=r"ring_dep_pool must be in \[4, INT32_MAX\]"):
+            RunConfig(platform="a2a3sim", ring_dep_pool=bad)
+
```
[the remark this answers is not in the document]
Comment (lyfne123) [reply tests/ut/runtime/test_run_config.py:202]: Keeping these tests since the validation stays (see the reply on the `runner.py` thread). They now also cover non-integer (float/bool) rejection added in `25607032`.

**Review by Hzfengsy — APPROVED** [review]
LGTM. Per-task ring sizing overrides on RunConfig are cleanly validated (mirroring RuntimeEnv::validate), forwarded only when set, well-tested, and documented in both EN/ZH. CI temporarily disables the DeepSeek-V4 attention examples for a clearly-documented external bf16 regression.
